File: 1114-printinorder/csharp/csharp_leetcode_version.cs
using System.Threading;

public class Foo {

    public Foo() {
        
    }
    
    static readonly object _locker = new object();

    int _state = 1;


    public void First(Action printFirst) {

        printFirst();

        lock (_locker) 
        {
            _state = 2; 
            Monitor.PulseAll(_locker); 
        }

    }

    public void Second(Action printSecond) {
        lock (_locker)
        {
            while (!(_state == 2))
            {
                Monitor.Wait(_locker);
            }
        }
        printSecond();

        lock (_locker)
        {
            _state = 3;
            Monitor.PulseAll(_locker);
        }
    }

    public void Third(Action printThird) {
        lock (_locker)
        {
            while (!(_state == 3))
                {
                Monitor.Wait(_locker);
            }
        }
        printThird();
    }
}

File: 1115-foobar/csharp/csharp_leetcode_version.cpp
using System.Threading;
public class FooBar {
    private int n;

    public FooBar(int n) {
        this.n = n;
    }
    
    int _state = 0;

    public void Foo(Action printFoo) {
        
        for (int i = 0; i < n; i++) {
            
            lock (this)
            {
                while (!(_state == 0))
                {
                    Monitor.Wait(this);
                }
            }
            
        	printFoo();
            
            lock (this)
            {
                _state = 1;
                Monitor.PulseAll(this);
            }
        }
    }

    public void Bar(Action printBar) {
        
        for (int i = 0; i < n; i++) {
            
            lock (this)
            {
                while (!(_state == 1))
                {
                    Monitor.Wait(this);
                }
            }
            
        	printBar();
            
            lock (this)
            {
                _state = 0;
                Monitor.PulseAll(this);
            }
        }
    }
}